### ISSUE TITLE
Unexpected "empty" behaviour

### ISSUE THREAD

**Issue by borisovg** [stedolan/jq#873]
This could well be me misunderstanding the way `empty` works but this is unexpected:

```
borisov@deli:~ $ echo '{"key1": {"attr1": "foo"}, "key2": {"attr1": "foo", "attr2": "bar"}}' | jq '.[].attr2 |= if . then "bar" else empty end'                               
null
```

Expected something like this (but without .key1.attr2 set to "null"):

```
borisov@deli:~ $ echo '{"key1": {"attr1": "foo"}, "key2": {"attr1": "foo", "attr2": "bar"}}' | jq '.[].attr2 |= if . then "bar" else . end'
{
  "key1": {
    "attr1": "foo",
    "attr2": null
  },
  "key2": {
    "attr1": "foo",
    "attr2": "bar"
  }
}
```

jq version: jq-1.5rc2

**Comment by pkoppstein** [comment]
@borisovg - Glad to see you're using 1.5rc2!

Regarding `empty`, I think you'll understand what's going on more clearly if you consider that expressions such as `{"a": empty}` yield nothing (i.e., the empty stream).  In a sense, "empty" is like a black hole.  It is certainly more like a black hole than a value.  Anyway, to remove a key from an object, you could use del/1, e.g.
```
$ jq -n -c '{"a":1, "b": 2} | del(.a)'
{"b":2}
```

**Comment by dtolnay** [comment]
Or update the parent object:

    echo '{"key1": {"attr1": "foo"}, "key2": {"attr1": "foo", "attr2": "bar"}}' |
        jq '.[] |= if .attr2 then (.attr2 = "bax") else . end'

**Comment by nicowilliams** [comment]
Here's a better demo:

    $ echo 1 2 3 | ./jq -cn 'reduce inputs as $i ([]; if $i==0 then empty else .+[$i] end)'
    [1,2,3]
    $ echo 1 2 3 0 | ./jq -cn 'reduce inputs as $i ([]; if $i==0 then empty else .+[$i] end)'
    null

Now, what should reduce do in this case?

Options:

 - all of reduce should `empty`
 - reduce should act as if the update expression ended in `//.`
 - raise an error
 - something else?  what?

**Comment by nicowilliams** [comment]
And of course, the reason this is about `reduce` is that `|=` uses `reduce` under the covers.

Here's a version of `_modify` that makes `|=` work as expected:

    $ jq -n 'def _modify(paths; update): reduce path(paths) as $p (.; setpath($p; getpath($p) | update) // .); {key1:{attr1:"foo"},key2:{attr1:"bar"}}|.[].attr2 |= if . then "bar" else empty end'
    {
      "key1": {
        "attr1": "foo"
      },
      "key2": {
        "attr1": "bar"
      }
    }

**Comment by nicowilliams** [comment]
Until we know what is the best fix, this should target 1.6.

I'm partial to fixing this in `_modify` and `_assign` as in the preceding reply, and leaving `reduce` alone in 1.5.  I'm OK with 1.5 having the update expression of `reduce` (and `foreach`) have undefined semantics if it backtracks, though we may want to document this.

**Comment by nicowilliams** [comment]
The workaround for `|=` are `... else . end` (per @dtolnay).

**Comment by nicowilliams** [comment]
Opinions as to how `reduce` should handle backtracking from update expressions?

**Comment by pkoppstein** [comment]
@nicowilliams asked:
> Opinions ...

`reduce` is fine as is, as the behavior and semantics are easily understood.  Furthermore it would be weird to make a special case of "reduce" (or are you proposing that "foreach" would change too?). 
In any case, changing reduce at this late stage of 1.5 development would be risky or needlessly delay 1.5.  And then there are backward compatibility issues.

By the way, it appears you missed one of the points raised in @borisovg's message.  He explicitly mentioned "if . then "bar" else . end".  

Anyway, one of the issues his post raises is whether there should be a way to use '=' (or '|=') to set a slot to some value if the slot exists, and otherwise do nothing. I believe he was expecting it would be possible (or would like to be able) to write something like .a //= nada' so that, for example, '{} | .a //= nada' would yield {}.

**Comment by nicowilliams** [comment]
Hmmm, yes, maybe I misread.  I don't think there's a way to do what's requested, but the change to `_modify` that I suggested allows it.

**Comment by borisovg** [comment]
@dtolnay: Thank you for your solution - it perfectly addresses my needs. :)

@nicowilliams and @pkoppstein: Just to clarify (although it is probably not necessary) that my problem with `if ... else . end` was that it unexpectedly modified "key1" to include a key1.attr2 which was previously not there. Because I (thought I had) limited scope to ".[].attr2" I expected that the black hole effect of `empty` would only affect key1.attr2 but it seems (for reasons that are still unknown to me) that the black hole is larger and more powerful than I had ever imagined! ;-)

**Comment by borisovg** [comment]
(Offtopic) BTW, I am loving 1.5 - the real scenario that the example above was derived from uses `gsub()` and it rocks - thank you!!

**Comment by nicowilliams** [comment]
BTW, we don't document what happens when the update expr of `reduce` or `foreach` backtracks.  We should first decide what we want that to be.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> Here's a version of _modify that makes |= work as expected

The change can easily be justified (because the current evaluation of `{} | .a |= empty` does not make much sense), but if one has such an expectation, wouldn't one also expect `{} | .a = empty` to evaluate to the same value?

**Comment by nicowilliams** [comment]
@pkoppstein Well, I'd expect `.a` to *not* get set in that case, not even to `null` if `.a` didn't exist already.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> we don't document what happens when the update expr of reduce or foreach backtracks.

The manual is written sparsely, which obviously allows for mischievous readings, but my reading of the documentation-by-example for "reduce" is that it already covers the case when the "expr" is empty:

> For each result that <the generator> produces, <expr> is run to accumulate a running total ...

As for the merits of changing the semantics, I would be interested to know whether anyone has existing code that would have to be changed.

**Comment by nicowilliams** [comment]
The "... is run to accumulate a running total ..." part is very specific to the example.  The subject is not touched at all.  I don't see any dispositive text there.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> Well, I'd expect .a to not get set in that case

Yes, but the question is: since '{} | .a = X' and '{} | .a |= X' evaluate to the same value when (null|X) evaluates to X, why should setting X to empty be an exception?  (That is, (null|empty) and (empty) both produce the empty stream.)

**Comment by nicowilliams** [comment]
`empty` -> backtrack.  There's no "empty value" as such.  So I expect `=` and `|=` to behave exactly the same as to RHS backtracking.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> So I expect = and |= to behave exactly the same as to RHS backtracking.

Yes, that's the point, as they don't. (With or without the change to _modify.)

For example, in 1.5rc2: `{} | .a = empty` yields the empty stream, but with |= the result is (weirdly) null.

A reasonable expectation would be that there is an expression, let's call it `nada`, such that:
```
 {"a": nada} == ({} | .a = nada) == ({} | .a |= nada) == {}
```

**Comment by nicowilliams** [comment]
No, there will never be a `nada`.  Using `empty` in the RHS of `=` and `|=` should work as I proposed earlier.

**Comment by nicowilliams** [comment]
The fix:

    def _assign(paths; value): (value as $v | reduce path(paths) as $p (.; setpath($p; $v)))//.;
    def _modify(paths; update): reduce path(paths) as $p (.; setpath($p; getpath($p) | update) // .);

in `builtin.c`.

We can do that in 1.5 and decide what `reduce` and `foreach` should do with backtracking from the reduction state update expression.

**Comment by nicowilliams** [comment]
From the OP's example:

    $ echo '{"key1": {"attr1": "foo"}, "key2": {"attr1": "foo", "attr2": "bar"}}' | jq 'def _modify(paths; update): reduce path(paths) as $p (.; setpath($p; getpath($p) | update) // .); .[].attr2 |= if . then "blah" else empty end'
    {
      "key1": {
        "attr1": "foo"
      },
      "key2": {
        "attr1": "foo",
        "attr2": "blah"
      }
    }
    $ 

I quite like it.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> The fix:

Yes! Modifying *both* _assign and _modify is the key!

**Comment by nicowilliams** [comment]
Eh, I never said "only fix _modify"...

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> there will never be a nada. 

Well, with the changes to _assign and _modify, you're about 95% there already.  In fact, since {"a": x} should (presumably) be the same as `{} | .a = x`, maybe anything else is a bug!

**Comment by nicowilliams** [comment]
Well, `{...}` doesn't function like `[...]`.  Currently `{a:empty}` yields nothing, not `{}`, whereas `[empty]` _of course_ yields `[]`.  It'd be nice to allow `{a:empty}` to mean `{}`, but... I wouldn't hold my breath :)

**Comment by nicowilliams** [comment]
Occasionally I've wanted `{exp}` to mean something like "collect key/value pairs output by exp into an object", something like:

    reduce exp as $e ([{}, null];
        if .[1]!=null then [.[0] + {.[1]:$e}]
        elif $e|type=="array" then [.[0] + {$e[0]:$e[1]}, null]
        else [., $e] end)

And that might be worth having as a function.  But I think that `{exp}` would be a bit too magical as far as syntax goes.

**Comment by nicowilliams** [comment]
I like this:

    def collect_object(exp): reduce (exp,null) as $e ([{}];
        if .[1]!=null then [.[0] + ([{key:.[1],value:$e}]|from_entries)]
        elif $e|type=="array" then [.[0] + ([{key:$e[0],value:$e[1]}]|from_entries), null]
        elif $e|type=="object" then [.[0] + $e, null]
        else [.[0], $e]
        end)|.[0];

**Comment by pkoppstein** [comment]
@nicowilliams - I believe that using the form "X // ." is incorrect here (or at least insufficient).  Perhaps `[X, .][0]` is closer to what is needed?

**Comment by nicowilliams** [comment]
It needs more work, yes.

**Comment by dtolnay** [comment]
`[X,.][0]` is also not correct because we want to keep this behavior:

    $ jq -n '{} | .x |= (1, 2)'
    {
      "x": 2
    }

**Comment by pkoppstein** [comment]
So is `[.,X][-1]` what is needed?

**Comment by nicowilliams** [comment]
We really need something like `//`, but without the conditional checking the value produced.  But even that's not good enough, since if the RHS produces at least one value we're bound to backtrack and that's OK.  This is starting to get too complicated, and it's easier to address it in the docs.

**Comment by pkoppstein** [comment]
@nicowilliams wrote:
> We really need something like // ...

Yes, I never understood why // had such apples-and-oranges semantics, but we're stuck with that. 

Given that currently '{} | .a = (1,2,3)' produces a stream of three values, the argument that '{} | .a = empty' should produce a stream of 0 values (i.e. empty) is quite strong.  

Is that the documentation you had in mind?

**Comment by nicowilliams** [comment]
Yes, I think that's right.  It's surprising, but it's correct.  `|=` works differently (`{b:2}|.a|=(1,2,3)` outputs `{b:2,a:3}`), and it's the one that is most surprising (`{b:2}|.a|=empty` produces `null`, but one might expect it to produce `{b:2}`).

**Comment by pkoppstein** [comment]
Assuming for the moment that we want to adhere to the principle:
```
the length of the stream (JSON | .[key] = STREAM) should be the length of STREAM # (*)
```
then I think we should focus on revamping |= to conform with the same rule.  Certainly, I don't see any good reason for wanting `{}|.a|=empty` => null.

So the bigger question is whether we want (*) above.  I doubt it's essential or even greatly used, and = and |= when taken together are currently weird enough to justify some clear-headed house-cleaning.  Since the main purpose of = and |= is to make things convenient and somewhat familiar to non-jq-programmers, I'd say: scrap (*) and focus on bringing | and |= into alignment with each other.

**Comment by nicowilliams** [comment]
`LHS ASSIGNMENT-OPERATOR RHS` should set all the paths named by `LHS` in `.` to whatever the `RHS` says to.  If there's many values of `RHS` they'll be produced as a stream for each path output/matched by `LHS`.  For many paths matched by `LHS` the useful thing to do is to reduce, and if we do that then it's not possible to do a cartesian product of reduction x the RHS values...

IMO `=` is broken.  Its definition should be more like `|=`, not the other way around.

I grant that the `|=` approach means that only the last value output by the `RHS` will take and be visible.  But again, I think the reduction aspect of `|=` is what is useful about it.

**Comment by pkoppstein** [comment]
I think we're agreed about | and |= w.r.t. reduction-style semantics, and with respect to their uniformity.

That still leaves:

a) Do we want `{a: STREAM}` to continue to produce a stream of ([STREAM]|length) items?

b) Do we want `{} | .a |= empty` to evaluate to {} or null?

Note that currently:
```
$ jq -n ' reduce empty as $i ({}; empty)
{}  # expected
$ jq -n ' reduce (1,2) as $i ({}; empty)
null # unexpected?
```

**Comment by nicowilliams** [comment]
I think we're going to be very conservative here and give at least a minor release's worth of warning before we change the semantics of any of these things.

Thinking about it, `=`'s semantics are useful (`null|(.a,.b,.c)=(0,1,2)` produces three objects whose "a", "b", and "c" keys have 0, then 1, then 2 as their values), and there's no way to get anything like them for `|=` because of its update semantics.  So I'm willing to leave that as-is.  Ditto `{key: EXP}`.

What's left to discuss?  The backtracking semantics for reduce/foreach state update expressions?  Maybe that's best left as-is too.

A utility function similar to `//`, but without taking the alternative when the LHS produces a false-ish value would be nice.  For some reason the simplest implementation I can think for it is something like `def foo(exp; alt): label $foo | foreach ((exp|[.]),null) as $e ([false]; if $e|length == 1 then [true, $e[0]] elif .[0] then break $foo else . end; if .[0] then .[1] else empty end), alt;`.

**Comment by nicowilliams** [comment]
Fixed by bd7b48c.  See also #1314.

`empty` in `|=` RHS now deletes the value at the LHS.
